yarn-project/p2p/src/service/libp2p_service.ts

```diff
@@ -0,0 +1,385 @@
+      // Propagation would need to filter and send custom tx set per peer
+      for (const tx of txs) {
+        await this.processTxFromPeer(tx, peerId);
+      }
+    } catch (err) {
+      this.logger(`Failed to process pooled transactions message`, err);
+    }
+  }
+
+  private async processTxFromPeer(tx: Tx, peerId: PeerId): Promise<void> {
+    const txHash = await tx.getTxHash();
+    const txHashString = txHash.toString();
+    this.knownTxLookup.addPeerForTx(peerId, txHashString);
+    this.logger(`Received tx ${txHashString} from peer ${peerId.toString()}`);
+    await this.txPool.addTxs([tx]);
+    this.propagateTx(tx);
+  }
+
+  private async sendTxToPeers(tx: Tx) {
+    const txs = createTransactionsMessage([tx]);
+    const payload = new Uint8Array(txs);
+    const peers = this.getTxPeers();
+    const txHash = await tx.getTxHash();
+    const txHashString = txHash.toString();
+    for (const peer of peers) {
+      try {
+        if (this.knownTxLookup.hasPeerSeenTx(peer, txHashString)) {
+          this.logger(`Not sending tx ${txHashString} to peer ${peer.toString()} as they have already seen it`);
+          continue;
+        }
+        this.logger(`Sending tx ${txHashString} to peer ${peer.toString()}`);
+        await this.sendRawMessageToPeer(payload, peer);
+        this.knownTxLookup.addPeerForTx(peer, txHashString);
+      } catch (err) {
+        this.logger(`Failed to send txs to peer ${peer.toString()}`, err);
+        continue;
+      }
+    }
+  }
+
+  private async sendTxHashesMessageToPeer(peer: PeerId) {
+    try {
+      const hashes = this.txPool.getAllTxHashes();
+      if (!hashes.length) {
+        return;
+      }
+      const message = createTransactionHashesMessage(hashes);
+      await this.sendRawMessageToPeer(new Uint8Array(message), peer);
+    } catch (err) {
+      this.logger(`Failed to send tx hashes to peer ${peer.toString()}`, err);
+    }
+  }
+
+  private async sendGetTransactionsMessageToPeer(hashes: TxHash[], peer: PeerId) {
+    try {
+      const message = createGetTransactionsRequestMessage(hashes);
+      await this.sendRawMessageToPeer(new Uint8Array(message), peer);
+    } catch (err) {
+      this.logger(`Failed to send tx request to peer ${peer.toString()}`, err);
+    }
+  }
+
+  private async sendTransactionsMessageToPeer(txs: Tx[], peer: PeerId) {
+    // don't filter out any transactions based on what we think the peer has seen,
+    // we have been explicitly asked for these transactions
+    const message = createTransactionsMessage(txs);
+    await this.sendRawMessageToPeer(message, peer);
+    for (const tx of txs) {
+      const hash = await tx.getTxHash();
+      this.knownTxLookup.addPeerForTx(peer, hash.toString());
+    }
+  }
+
+  private async sendRawMessageToPeer(message: Uint8Array, peer: PeerId) {
+    const stream = await this.node.dialProtocol(peer, this.protocolId);
+    await pipe([message], stream);
+    stream.close();
+  }
```

Comment:
Out of curiosity, I checked what lodestar does for sending a message. It seems we still need to bring in a lot more boilerplate in here...

https://github.com/ChainSafe/lodestar/blob/722ccb69a1404f7fbf2ccda32b29c7bf9420a630/packages/reqresp/src/request/index.ts

Reply:
That page is really useful. It's part of their implementation of the ReqResp pattern used in Ethereum's consensus. Most of the code in that file looks to me to be handling multiple protocol ids, timeout enforcement, chunking of data etc. We will need to consider that sort of thing I am sure for a 'production' version. For now all I've done is open a channel throw a message down it and close it again.